File: Tree/257. Binary Tree Paths/binaryTreePaths.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
var binaryTreePaths = function (root) {
    let paths = [];
    dfs(root, "", paths);
    return paths;

    function dfs(root, path, paths) {
        if (!root) return;

        if (!root.left && !root.right) {
            path += root.val;
            paths.push(path);
        }
        // add the path after the if statement 
        // so i don't need to slice the string 
        // e.g. path.slice(0, -2)
        path += root.val + "->";

        dfs(root.left, path, paths);
        dfs(root.right, path, paths);
    }
};

File: Structures & Algo/B. Tree/Backtracking/257. Binary Tree Paths/binaryTreePaths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        paths = []
        
        def traverse(path, root):
            if not root:
                return
            if not root.left and not root.right:
                path.append(str(root.val))
                paths.append("->".join(path))
                path.pop()
                return
            
            path.append(str(root.val))
            traverse(path, root.left)
            traverse(path, root.right)
            path.pop()
        
        traverse([], root)
        return paths